File: package/main/src/String/padStart.ts
/**
 * Pads the start of a string with another string until the target length is reached
 *
 * @param {string} string_ - The original string to pad
 * @param {number} targetLength - The target length after padding
 * @param {string} padString - The string to use for padding
 * @returns {string} The padded string
 * @example
 * ```typescript
 * padStart("123", 5, "0");     // Returns: "00123"
 * padStart("abc", 8, "def");   // Returns: "defdeabc"
 * ```
 */
export const padStart = (
  string_: string,
  targetLength: number,
  padString: string,
): string => {
  // Return original string if padString is empty or string is already long enough
  if (padString === "" || string_.length >= targetLength) {
    return string_;
  }

  let padding = "";
  const paddingLength = targetLength - string_.length;

  // Build padding by repeating padString
  while (padding.length < paddingLength) {
    padding += padString;
  }

  // Trim padding to exact length needed and concatenate with original string
  return padding.slice(0, paddingLength) + string_;
};

File: package/umt_python/src/pad_start.py
def pad_start(string_: str, target_length: int, pad_string: str) -> str:
    """
    Pads the start of a string with another string until the target length is reached.

    Args:
        string_: The original string to pad.
        target_length: The target length after padding.
        pad_string: The string to use for padding.

    Returns:
        The padded string.

    Raises:
        ValueError: If pad_string is empty.

    Example:
        >>> pad_start("123", 5, "0")
        '00123'
        >>> pad_start("abc", 8, "def")
        'defdeabc'
        >>> pad_start("world", 3, "!") # target_length is less than string length
        'world'
    """
    if not pad_string:
        raise ValueError("pad_string cannot be empty")

    if len(string_) >= target_length:
        return string_

    padding = ""
    padding_length = target_length - len(string_)

    while len(padding) < padding_length:
        padding += pad_string
    
    return padding[:padding_length] + string_
